[JENKINS-34389] Fixed handling of dynamic axis

bboehmke opened the pull request:
Fixed handling of concurrent builds with different axis by adding a mutex and deep copy the AxisList.

https://issues.jenkins-ci.org/browse/JENKINS-34389

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -207,6 +214,24 @@ public MatrixProject(ItemGroup parent, String name) {
+    public void buildLock(){
+        // create lock instance if not exist
+        if (buildLock == null) {
+            buildLock = new ReentrantReadWriteLock();
```
You will run into problems when 2 threads call this at the same time - they will use different lock instance and operate in parallel anyway. The first one that completed frees the lock used by latter thread for third one to enter the critical section while 2nd can still be in progress.

olivergondza commented:
I really do not like this approach. The idea of exposing project lock for everybody to use (and potentially  fail to release - you are not releasing in finally) is terribly fragile.

Please attach a reproducer test so we know how the problem manifests exactly. Do we really need to synchronize `#rebuildConfigurations()` and the aggregator listing? Once we know what needs to be synchronized, we need to move that into single method on project that will be responsible for consistency of its results in concurrent situations.

bboehmke commented:
I know the fix is not very well and was only created to get our setup with dynamic matrix axis working.

The important thing that should be synchronized is the creation of the axis. If the doRun function is called from different threads with different axis the first one is overwrriten and the jobs can not be accessed anymore (also it looks strange in the GUI).

In the issue is an attachment of a job setting that should reproduce the behavior.

olivergondza commented:
> In the issue is an attachment of a job setting that should reproduce the behavior.

I have seen that. What I had in mind is a reproducing unit-test.

bboehmke commented:
I am not sure how to create how to create a module test for this case. 
I can try it but I have never create a unit test for a java application.

bboehmke commented:
I have added a test that can reproduce the problem with the dynamic axis (I have added a simple axis class for this).
I also improved the implementation of the mutex based on your remarks.

bboehmke commented:
@olivergondza Is there any progress for this pull request or is there something I should do to get this merged?

olivergondza commented on src/main/java/hudson/matrix/MatrixBuild.java:
```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
-            activeConfigurations = p.rebuildConfigurations(this);
+            try {
+                // lock the project
+                p.buildLock();
```
- The lock should not be used from the outside. The logic of `listUpAggregators` belongs to `MatrixProject` where the locking should take place.
- What operation can actually race with this? I would say it is configuration update and perhaps rebuilding configurations but I do not see them synchronized over the same lock.

bboehmke replied:
Your right. I think only the `rebuildConfigurations` causes the problem and has no impact on `listUpAggregators`. They should be moved outside the lock.

If the lock should only be used inside `MatrixProject` a new function must be created that returns the active configuration and the deep copy of the axes. Because such a function also would require the build context for `rebuildConfigurations` I am not sure if it is a good solution

olivergondza replied:
I would still go that way.

olivergondza commented on src/main/java/hudson/matrix/MatrixBuild.java:
```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
-            listUpAggregators(p.getProperties().values());
-            listUpAggregators(p.getBuildWrappers().values());
+                // deep copy the axes
+                axes = (AxisList) Jenkins.XSTREAM.fromXML(Jenkins.XSTREAM.toXML(p.getAxes()));
```
`new AxisList(p.getAxes());`

bboehmke replied:
I do not think this would work because a deep copy is required and the axis do not have a copy constructor or similar (or at least I have not found them)

olivergondza replied:
It is a copy constructor.

bboehmke replied:
Yes but it creates only a copy of the list not of the axes itself. Axes instances are keeped

olivergondza replied:
And those instances are being modified?

bboehmke replied:
Yes that is the problem. The axisValues are cached and will be regenerated by `rebuildConfigurations`.

Btw the Unit test fails if I use `axes = new AxisList(p.getAxes());`

olivergondza replied:
> Yes that is the problem. The axisValues are cached and will be regenerated by rebuildConfigurations.

You are right. This seems to be specific to dynamic axis plugin (and your reproducer). I see `Axis#rebuild` is ill-designed in this regard, should return a copy of whole axis so deep copying can be implemented. I see no better way of doing so.

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -199,14 +201,35 @@
+    /**
+     * Lock to prevent project changes on different build at the same time
+     */
+    private ReadWriteLock buildLock;
```
iff we need this, it should be transient so it is not persisted and initialized in `readResolve` so old configuration have it configured. As implemented it would cause NPEs when old jobs are build.

olivergondza replied:
Yes, and reinitialize that in `readResolve` method (constructors and initializers are not called for deserialization).

bboehmke replied:
OK and because the function is not already present I simple add:
```
protected Object readResolve() {
    if (buildLock == null) {
        buildLock = new ReentrantReadWriteLock();
    }
    return this;
}
```
(as described [here](https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility))

olivergondza replied:
`buildLock = new ReentrantReadWriteLock();` should be enough as transient fields will be null.

olivergondza commented on src/test/java/hudson/matrix/MatrixTest.java:
```diff
@@ -135,4 +141,32 @@ private static void expectRejection(MatrixProject project, String combinationFil
+
+
+    @Issue("JENKINS-34389")
+    @Test public void axisValuesChanged() throws Exception {
```
Confirmed to reproduce the issue. Thanks

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -591,7 +603,7 @@ public boolean accept(File child) {
      *      build in progress. Otherwise this value is null (for example, when Jenkins is booting up.)
      */
-    /*package*/ Set<MatrixConfiguration> rebuildConfigurations(MatrixBuildExecution context) throws IOException {
+    private Set<MatrixConfiguration> rebuildConfigurations(MatrixBuildExecution context) throws IOException {
```
This is unsafe change.

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -660,6 +672,34 @@ public Combination apply(@Nullable List<String> strings) {
+    /*package*/ RunConfiguration getRunConfiguration(MatrixBuildExecution context) throws IOException {
+        RunConfiguration runConfig = new RunConfiguration();
+        try {
+            buildLock.writeLock().lock();
```
As I understand this provides week concurrency guarantee as it protects 2 concurrent dynamic builds but not dynamic build and, say, config update. Is that ok for you?

bboehmke replied:
Maybe it is possible if the job is changed in the moment a build starts but because the way the axes are build I see no better solution.

I think the point that the axes are only rebuild if the build context is given may also helps to reduce the collisions.
Actual we are using the first bad implementation in our productive Jenkins without any problems since a half year.

olivergondza replied:
I am fine with that as the problem existed for a long time.

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -660,6 +672,34 @@ public Combination apply(@Nullable List<String> strings) {
+    /**
+     * Configuration for matrix build
+     */
+    public static class RunConfiguration {
```
This can be `/*package*/` as well.

olivergondza commented on src/main/java/hudson/matrix/MatrixProject.java:
```diff
@@ -207,6 +214,11 @@ public MatrixProject(ItemGroup parent, String name) {
     }
 
+    protected Object readResolve() {
+        buildLock = new ReentrantReadWriteLock();
```
`ReentrantLock` would suffice since we do not protest writing.

bboehmke replied:
Yes the normal lock should be enough

olivergondza commented:
It looks good to me otherwise.

olivergondza commented:
Merging, thanks!

Softmuppen commented:
Hi,

We receive the following error on our parent jobs. We are running multiple jobs in parallel

```
java.lang.IllegalMonitorStateException
	at java.util.concurrent.locks.ReentrantLock$Sync.tryRelease(ReentrantLock.java:151)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.release(AbstractQueuedSynchronizer.java:1261)
	at java.util.concurrent.locks.ReentrantLock.unlock(ReentrantLock.java:457)
	at hudson.matrix.MatrixProject.getRunConfiguration(MatrixProject.java:698)
	at hudson.matrix.MatrixBuild$MatrixBuildExecution.doRun(MatrixBuild.java:359)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
	at hudson.model.Run.execute(Run.java:1720)
	at hudson.matrix.MatrixBuild.run(MatrixBuild.java:313)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:404)
```

How can we solve this?

olivergondza commented:
@Softmuppen

Strange, the thread does not seem to own the lock it is trying to `unlock()`. The most compelling reason I can see is the `lock()` has thrown never acquiring the lock so `finally` branch fails hiding the original exception. It is actually a mistake to `lock` inside of the `try-catch`. Adjusting in 3966482d0272f2226b76767a62ede3266c2cf655, but chances are this only masks the true problem.

How often do you see this. Can you reproduce? Are concurrent builds in play?

Softmuppen commented:
@olivergondza 

Yes, we are running concurrent jobs. We experienced that the fault was intermittent, around 5% of our builds failed to this lock/unlock error. We had to roll back the plugin to 1.7 (the one we used previously).

olivergondza commented:
@Softmuppen, I have released the improvement in 1.12.

Softmuppen commented:
@olivergondza

Thanks for the fast update!
We will upgrade the plugin version to 1.12 first thing tomorrow morning. I will get back to you with status.

Softmuppen commented:
We updated to 1.12 this morning and it did not take long before this stacktrace appeared. We are rolling back to 1.7 for now. 

@olivergondza Any ideas?

```
10:37:33 FATAL: null
10:37:33 java.lang.IllegalMonitorStateException
10:37:33 	at java.util.concurrent.locks.ReentrantLock$Sync.tryRelease(ReentrantLock.java:151)
10:37:33 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.release(AbstractQueuedSynchronizer.java:1261)
10:37:33 	at java.util.concurrent.locks.ReentrantLock.unlock(ReentrantLock.java:457)
10:37:33 	at hudson.matrix.MatrixProject.getRunConfiguration(MatrixProject.java:696)
10:37:33 	at hudson.matrix.MatrixBuild$MatrixBuildExecution.doRun(MatrixBuild.java:361)
10:37:33 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
10:37:33 	at hudson.model.Run.execute(Run.java:1720)
10:37:33 	at hudson.matrix.MatrixBuild.run(MatrixBuild.java:314)
10:37:33 	at hudson.model.ResourceController.execute(ResourceController.java:98)
10:37:33 	at hudson.model.Executor.run(Executor.java:404)
```

olivergondza commented:
@Softmuppen, I am struggling to reproduce this in unittest. Can you share your job `config.xml`? Also, how many builds in parallel are you running?

Softmuppen commented:
@olivergondza

We are running around 10 matrix builds of this `config.xml` at the same time, using dynamic axis, these matrices are different each time. Due to company privacy we cant share the whole `config.xml` , can you specify which parts you are interested in?

We were looking at the ReentrantLock class java doc [example](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html), and we noticed that they set their variable to `final`, is there a reason for yours not to be? Could it be an issue if `MatrixProject:readResolve()` is called before we try to release the lock, and is called from another thread? We are just speculating!

olivergondza commented:
> is there a reason for yours not to be [`final`]?

I would love to make it final but the braindead Java attitude to serialization constructors / `final` field initialization does not allow that*.

> can you specify which parts you are interested in?

Feel free to skip the builders, post builds, SCMs, etc. Or, you can reproduce internally a snapshot build I provide you.

> Could it be an issue if MatrixProject:readResolve() is called before we try to release the lock

Might be but the "real" invocation of that method should happen long before matrix build is run. I see no reason why anything would call it again. But the fact the lock instance is either replaced or its state is tempered with by other thread while in use is my working hypothesis.

*) Transient fields are not persisted and therefore `null` after loaded (why not calling initializes for `transient` fields during initialization?). So it needs to be fixed in serialization pseudo-constructor such as `readResolve` (constructors/initializers are not invoked) but it is not allowed to modify final fields is those are not considered constructors (why, java, why?).

Softmuppen commented:
@olivergondza 

We will try to reproduce the issue with a minimal config.xml then send it to you

Dergons commented:
Hello!

I'm colleagues with @Softmuppen and have been working on reproducing the fault we saw in our live environment. Pardon the extensive radio silence, we've made several failed attempts at this previously and hadn't been successful until now.

It seems to be a requirement that there are a multitude of nodes active and serving the chain of jobs, for the error to occur. Our test environment only uses a limited number of nodes, causing the fault to be absent, but by copying these, thus simulating the use of 12 nodes, we managed get it to reoccur.

I have attached configs for a chain of three jobs, a start job (IllegalMonitor_Start), the main matrix job (IllegalMonitor_Matrix) and a postbuild job (IllegalMonitor_Matrix_Cleanup), that reproduces the fault, at least in our environment. To produce the fault, setup a bunch of nodes with a bunch of executors, as mentioned we duplicated our existing nodes to simulate the use of 12 nodes (with 10 executors each), setup the job chain and spam-trigger builds of the start job. Typically the fault occurs within 10-30 builds, although sometimes I've had it first occurring at ~80-100 builds

 I also identified that it is the groovy postbuild action in the post-build job that causes the error, namely: 

`TopLevelItem upstreamJob = Jenkins.getInstance().getItem(upstreamJobName).doReload()`

If it is removed, the fault disappears.

Some observations: 

The fault seems sporadic. Any time it occurs it will be while triggering builds on the **same** node, for some time, and then the node where the fault occurs will switch. For example the fault might occur when triggering builds on node_1 for a few hours or a day, and then switch to occurring while building on node_7. 

There have also been situations where I've stripped the config of some features, saw that the fault disappeared, rolled back the config and again, piece by piece, stripping it to the same config as when the fault previously disappeared only to have the fault occur again. Some of this might be due to variance in the fault occurring but I seriously doubt all of it was.


[IllegalMonitor_Configs.zip](https://github.com/jenkinsci/matrix-project-plugin/files/2119298/IllegalMonitor_Configs.zip)

olivergondza commented:
@Dergons, please fine a new issue on JIRA so it does not get forgotten at the bottom of a PR closed over a year ago.

olivergondza commented:
@Dergons, please fine a new issue on JIRA so it does not get forgotten at the bottom of a PR closed over a year ago.
